istio-2-mixin/alerts.libsonnet

```diff
@@ -0,0 +1,163 @@
+        'for': '5m',
+        labels: {
+          severity: 'critical',
+        },
+        annotations: {
+          summary: 'There are a high number of HTTP request errors in the Istio system.',
+          description: |||
+            HTTP requests from Istio service {{$labels.source_canonical_service}} to service {{$labels.destination_canonical_service}} on cluster {{$labels.cluster}} have an error rate above {{ printf "%%.0f" $value }}%%, which is above the threshold of %(alertsCriticalHTTPRequestErrorPercentage)s%%.
+          ||| % this.config,
+        },
+      },
+      {
+        alert: 'IstioHighGRPCRequestErrorsCritical',
+        expr: |||
+          100 * sum without(connection_security_policy, destination_app, destination_canonical_revision, destination_service_name, destination_cluster, destination_principal, destination_service, destination_service_namespace, destination_version, destination_workload, destination_workload_namespace, grpc_response_status, instance, pod, reporter, request_protocol, response_code, response_flags, source_app, source_canonical_revision, source_cluster, source_principal, source_version, source_workload, source_workload_namespace) (increase(istio_requests_total{%(filteringSelector)s, %(reporterSourceFilter)s, %(grpcResponseStatusErrorFilter)s}[5m]))
+          /
+          clamp_min(sum without(connection_security_policy, destination_app, destination_canonical_revision, destination_service_name, destination_cluster, destination_principal, destination_service, destination_service_namespace, destination_version, destination_workload, destination_workload_namespace, grpc_response_status, instance, pod, reporter, request_protocol, response_code, response_flags, source_app, source_canonical_revision, source_cluster, source_principal, source_version, source_workload, source_workload_namespace) (increase(istio_requests_total{%(filteringSelector)s, %(reporterSourceFilter)s, %(grpcResponseStatusFilter)s}[5m])), 1) > %(alertsCriticalGRPCRequestErrorPercentage)s
+        ||| % this.config { reporterSourceFilter: reporterSourceFilter, grpcResponseStatusErrorFilter: grpcResponseStatusErrorFilter, grpcResponseStatusFilter: grpcResponseStatusFilter },
+        'for': '5m',
+        labels: {
+          severity: 'critical',
+        },
+        annotations: {
+          summary: 'There are a high number of GRPC request errors in the Istio system.',
+          description: |||
+            GRPC requests from Istio service {{$labels.source_canonical_service}} to service {{$labels.destination_canonical_service}} on cluster {{$labels.cluster}} have an error rate above {{ printf "%%.0f" $value }}%%, which is above the threshold of %(alertsCriticalGRPCRequestErrorPercentage)s%%.
+          ||| % this.config,
+        },
+      },
+      {
+        alert: 'IstioMetricsDown',
+        expr: |||
+          up{%(filteringSelector)s} == 0
```

Comment:
Will this trigger if the prometheus exporter is running on the host itself and it goes down? Or is this an unlikely situation given Istio architecture and they are separate?

Reply:
So here is my concern on a more clear example.  Say you have a VM instance with MySQL and a prometheus exporter on the same instance. Then if something wrong occurs with the instance, the metrics won't be reporting at all and the value is nothing, not 0 right?  

Looking back at it, I suppose this is not saying istio down, but is saying istio metrics down. So it is saying something is wrong with the prometheus exporter or configuration to scrape and emit successfully. Maybe it is fine then.